clang/include/clang/AST/Type.h

```diff
@@ -486,7 +486,13 @@ class Qualifiers {
   /// Returns true if the address space in these qualifiers is equal to or
   /// a superset of the address space in the argument qualifiers.
   bool isAddressSpaceSupersetOf(Qualifiers other) const {
-    return isAddressSpaceSupersetOf(getAddressSpace(), other.getAddressSpace());
+    return isAddressSpaceSupersetOf(getAddressSpace(),
+                                    other.getAddressSpace()) ||
+           (!hasAddressSpace() &&
+            (other.getAddressSpace() == LangAS::sycl_private ||
+             other.getAddressSpace() == LangAS::sycl_local ||
+             other.getAddressSpace() == LangAS::sycl_global ||
+             other.getAddressSpace() == LangAS::sycl_constant));
```

Comment:
Hmm, maybe it would be good to add a comment explaining why we allowed casting between "empty" address space and others. At least for casts to `sycl_constant` address spaces.

Reply:
> I don't think this casts should be allowed as they are not legal in OpenCL env.

Agree.